velox/common/memory/tests/StreamArenaTest.cpp

```diff
@@ -0,0 +1,162 @@
+       0,
+       largestClassPageSize * 5 + AllocationTraits::kPageSize * 3},
+      {{largestClassPageSize - 1,
+        1,
+        largestClassPageSize + 1,
+        largestClassPageSize / 2,
+        1,
+        AllocationTraits::kPageSize + 1,
+        AllocationTraits::kPageSize},
+       {largestClassPageSize - 1,
+        1,
+        largestClassPageSize + AllocationTraits::kPageSize,
+        largestClassPageSize / 2,
+        1,
+        AllocationTraits::kPageSize + 1,
+        AllocationTraits::kPageSize - 2},
+       largestClassPageSize + largestClassPageSize / 2 +
+           AllocationTraits::kPageSize * 2,
+       largestClassPageSize + AllocationTraits::kPageSize}};
+
+  for (const auto& testData : testSettings) {
+    SCOPED_TRACE(testData.debugString());
+    ASSERT_EQ(
+        testData.requestRangeSizes.size(), testData.expectedRangeSizes.size());
+
+    auto arena = newArena();
+    ByteRange range;
+    for (int i = 0; i < testData.requestRangeSizes.size(); ++i) {
+      arena->newRange(testData.requestRangeSizes[i], &range);
+      ASSERT_EQ(range.size, testData.expectedRangeSizes[i]) << range.toString();
+      ASSERT_EQ(range.position, 0);
+      ASSERT_TRUE(range.buffer != nullptr);
+    }
+    ASSERT_EQ(
+        testData.expectedContiguousAllocationSize +
+            testData.expectedNonContiguousAllocationSize,
+        arena->size());
+    const int allocateBytes =
+        AllocationTraits::pageBytes(mmapAllocator_->numAllocated());
+    const int contiguousBytes =
+        AllocationTraits::pageBytes(mmapAllocator_->numExternalMapped());
+    ASSERT_EQ(
+        testData.expectedNonContiguousAllocationSize,
+        allocateBytes - contiguousBytes);
+    ASSERT_EQ(testData.expectedContiguousAllocationSize, contiguousBytes);
+  }
+}
+
+TEST_F(StreamArenaTest, randomRange) {
+  const int numRanges = 30;
+  auto arena = newArena();
+  ByteRange range;
+  for (int i = 0; i < numRanges; ++i) {
+    if (folly::Random::oneIn(4)) {
+      const int requestSize =
+          1 + folly::Random::rand32() % (2 * AllocationTraits::kPageSize);
+      arena->newTinyRange(requestSize, &range);
+      ASSERT_EQ(range.size, requestSize);
+    } else if (folly::Random::oneIn(3)) {
+      const int requestSize =
+          AllocationTraits::pageBytes(pool_->largestSizeClass()) +
+          (folly::Random::rand32() % (4 << 20));
+      arena->newRange(requestSize, &range);
+      ASSERT_EQ(AllocationTraits::alignedPageBytes(requestSize), range.size);
+    } else {
+      const int requestSize =
+          1 + folly::Random::rand32() % pool_->largestSizeClass();
+      arena->newRange(requestSize, &range);
+      ASSERT_LE(range.size, AllocationTraits::alignedPageBytes(requestSize));
+    }
+    ASSERT_EQ(range.position, 0);
+    ASSERT_TRUE(range.buffer != nullptr);
+  }
+}
+
+TEST_F(StreamArenaTest, error) {
+  auto arena = newArena();
+  ByteRange range;
+  VELOX_ASSERT_THROW(arena->newTinyRange(0, &range), "");
```

Comment:
Maybe add an error message and test it here? It is hard to debug when there is no error message.